File: src/main/java/com/github/svegon/capi/event/render/tooltip/ItemTooltipListener.java
package com.github.svegon.capi.event.render.tooltip;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.item.TooltipContext;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public interface ItemTooltipListener {
    Event<ItemTooltipListener> EVENT = EventFactory.createArrayBacked(ItemTooltipListener.class,
            (stack, player, context, tooltip) -> {}, listeners -> (stack, player, context, tooltip) -> {
                for (ItemTooltipListener listener : listeners) {
                    listener.addItemTooltip(stack, player, context, tooltip);
                }
            });

    void addItemTooltip(ItemStack stack, @Nullable PlayerEntity player, TooltipContext context, List<Text> tooltip);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/tooltip/ItemTooltipListener.kt
package io.github.svegon.capi.event.render.tooltip

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.item.TooltipContext
import net.minecraft.text.Text
import java.util.function.Function

interface ItemTooltipListener {
    fun addItemTooltip(stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?, tooltip: List<Text?>?)

    companion object {
        @JvmField
        val EVENT: Event<ItemTooltipListener?> = EventFactory.createArrayBacked<ItemTooltipListener?>(
            ItemTooltipListener::class.java,
            ItemTooltipListener { stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?, tooltip: List<Text?>? -> },
            Function<Array<ItemTooltipListener?>, ItemTooltipListener?> { listeners: Array<ItemTooltipListener?>? ->
                ItemTooltipListener { stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?, tooltip: List<Text?>? ->
                    for (listener in listeners) {
                        listener.addItemTooltip(stack, player, context, tooltip)
                    }
                }
            })
    }
}
